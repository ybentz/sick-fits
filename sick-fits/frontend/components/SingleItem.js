import React from 'react'
import gql from 'graphql-tag'
import { useQuery } from '@apollo/react-hooks'
import styled from 'styled-components'
import Head from 'next/head'
import Error from './ErrorMessage'

const SingleItemStyles = styled.div`
  max-width: 1200px;
  min-height: 800px;
  margin: 2rem auto;
  box-shadow: ${(props) => props.theme.bs};
  display: grid;
  grid-auto-columns: 1fr;
  grid-auto-flow: column;
  img {
    width: 100%;
    height: 100%;
    object-fit: contain;
  }
  .details {
    margin: 3rem;
    font-size: 2rem;
  }
`

const SINGLE_ITEM_QUERY = gql`
  query SINGLE_ITEM_QUERY($id: ID!) {
    item(where: { id: $id }) {
      id
      title
      description
      largeImage
    }
  }
`

function SingleItem(props) {
  const { error, loading, data } = useQuery(SINGLE_ITEM_QUERY, {
    variables: { id: props.id },
  })

  if (error) return <Error error={error} />
  if (loading) return <p>Loading...</p>
  if (!data.item) return <p>No item found for {this.props.id}</p>

  const { item } = data
  return (
    <SingleItemStyles>
      <Head>
        <title>Sick Fits | {item.title}</title>
      </Head>
      <img src={item.largeImage} alt="Item image" />
      <div className="details">
        <h2>Viewing {item.title}</h2>
        <p>{item.description}</p>
      </div>
    </SingleItemStyles>
  )
}

export default SingleItem
export { SINGLE_ITEM_QUERY }
