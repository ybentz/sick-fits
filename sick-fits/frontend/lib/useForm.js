import { useState } from 'react'

function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial)

  const handleChange = (event) => {
    const { name, type, value } = event.target
    const val = type === 'number' ? parseFloat(value) : value
    setInputs({
      ...inputs,
      [name]: val,
    })
  }

  return {
    inputs,
    handleChange,
  }
}

export default useForm
