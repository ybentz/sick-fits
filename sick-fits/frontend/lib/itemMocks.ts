// import { MockedResponse } from '@apollo/react-testing'

import { fakeUser } from './testUtils'
import { SINGLE_ITEM_QUERY } from '../components/SingleItem'

class ItemMockBuilder implements Item {
  id: string
  description: string
  image: string
  largeImage: string
  price: number
  title: string
  user: User

  constructor() {
    this.id = '1234'
    this.description = 'Test description'
    this.image = 'test-dog-small.jpg'
    this.largeImage = 'test-dog.jpg'
    this.price = 5000
    this.title = 'Test Product Title'
    this.user = fakeUser()
  }
  withId(id: string) {
    this.id = id
    return this
  }
  withDescription(description: string) {
    this.description = description
    return this
  }
  withImage(url: string) {
    this.image = url
    return this
  }
  withLargeImage(url: string) {
    this.largeImage = url
    return this
  }
  withPrice(price: number) {
    this.price = price
    return this
  }
  withTitle(title: string) {
    this.title = title
    return this
  }
  withUser(user: User) {
    this.user = user
    return this
  }
  build(): Item {
    return {
      description: this.description,
      id: this.id,
      image: this.image,
      largeImage: this.largeImage,
      price: this.price,
      title: this.title,
      user: this.user,
    }
  }
}

class SingleItemQueryMockBuilder extends ItemMockBuilder {
  errorMessage: string

  withError(message: string) {
    this.errorMessage = message
    return this
  }

  // Should be returning type `MockedResponse` instead of `any but that complicates things
  // because TS doesn't handle the overload with same parameters signature this way
  build(): any {
    const request = { query: SINGLE_ITEM_QUERY, variables: { id: this.id } }
    if (this.errorMessage) {
      return {
        request,
        result: {
          errors: [{ message: this.errorMessage }],
        },
      }
    }

    return {
      request,
      result: {
        data: {
          item: {
            description: this.description,
            id: this.id,
            image: this.image,
            largeImage: this.largeImage,
            price: this.price,
            title: this.title,
            user: this.user,
          },
        },
      },
    }
  }
}

export { ItemMockBuilder, SingleItemQueryMockBuilder }
