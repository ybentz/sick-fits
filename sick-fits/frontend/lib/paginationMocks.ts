import { PAGINATION_QUERY } from '../components/Pagination'

class PaginationQueryMockBuilder {
  count: number
  errorMessage: string

  constructor() {
    this.count = 1
  }
  withItemCount(count: number) {
    this.count = count
    return this
  }
  withError(message: string) {
    this.errorMessage = message
    return this
  }

  build() {
    const request = { query: PAGINATION_QUERY }
    if (this.errorMessage) {
      return {
        request,
        result: {
          errors: [{ message: this.errorMessage }],
        },
      }
    }

    return {
      request,
      result: {
        data: {
          itemsConnection: {
            aggregate: {
              count: this.count,
            },
          },
        },
      },
    }
  }
}

export { PaginationQueryMockBuilder }
