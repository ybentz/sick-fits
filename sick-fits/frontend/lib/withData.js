import { InMemoryCache } from 'apollo-cache-inmemory'
import { ApolloClient } from 'apollo-client'
import { ApolloLink, Observable } from 'apollo-link'
import { onError } from 'apollo-link-error'
import { HttpLink } from 'apollo-link-http'
import withApollo from 'next-with-apollo'

import { LOCAL_STATE_QUERY } from '../components/Cart'
import { endpoint } from '../config'

const cache = new InMemoryCache()

async function request(operation, options) {
  operation.setContext({
    fetchOptions: {
      credentials: 'include',
    },
    headers: options.headers,
  })
}

function requestLinkWithOptions(options) {
  return new ApolloLink(
    (operation, forward) =>
      new Observable((observer) => {
        let handle
        Promise.resolve(operation)
          .then((oper) => request(oper, options))
          .then(() => {
            handle = forward(operation).subscribe({
              next: observer.next.bind(observer),
              error: observer.error.bind(observer),
              complete: observer.complete.bind(observer),
            })
          })
          .catch(observer.error.bind(observer))
        return () => {
          if (handle) handle.unsubscribe()
        }
      })
  )
}

function createClient({ headers }) {
  return new ApolloClient({
    cache,
    link: ApolloLink.from([
      onError(({ graphQLErrors, networkError }) => {
        if (graphQLErrors)
          // Handle/log graphQL errors - just console logging for now
          graphQLErrors.forEach(({ message, locations, path }) =>
            console.log(
              `[GraphQL error]: Message: ${message}, Location: ${locations}, Path: ${path}`
            )
          )
        // Handle/log network errors - just console logging for now
        if (networkError) console.log(`[Network error]: ${networkError}`)
      }),
      requestLinkWithOptions({ headers }),
      new HttpLink({
        uri: process.env.NODE_ENV === 'development' ? endpoint : endpoint,
        credentials: 'include',
      }),
    ]),
    resolvers: {
      Mutation: {
        // first argument unknown and unused lol
        toggleCart(_, variables, { cache }) {
          // read cartOpen value from cache
          const { cartOpen } = cache.readQuery({
            query: LOCAL_STATE_QUERY,
          })
          const data = {
            data: { cartOpen: !cartOpen },
          }
          cache.writeData(data)
          return data
        },
      },
    },
  })
}

// Add default local state
cache.writeData({
  data: {
    cartOpen: false,
  },
})

export default withApollo(createClient)
