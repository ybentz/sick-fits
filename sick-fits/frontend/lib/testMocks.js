import { CURRENT_USER_QUERY } from '../components/User'
import { fakeUser, fakeCartItem } from '../lib/testUtils'

const notSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: null } },
  },
]

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { me: fakeUser() } },
  },
]

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        me: {
          ...fakeUser(),
          cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
        },
      },
    },
  },
]

export { notSignedInMocks, signedInMocks, signedInMocksWithCartItems }
