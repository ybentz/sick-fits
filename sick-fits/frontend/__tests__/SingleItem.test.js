import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem'
import { fakeItem } from '../lib/testUtils'

describe('<SingleItem/>', () => {
  it('should render', async () => {
    const mocks = [
      {
        // when this request is made
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this mock
        result: {
          data: {
            item: fakeItem(),
          },
        },
      },
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SingleItem id="123" />
      </MockedProvider>
    )
    expect(wrapper.text()).toBe('Loading...')
    await wait()
    wrapper.update()
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot()
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot()
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot()
  })

  it('should render error', async () => {
    const errorText = 'Items not found'
    const mocks = [
      {
        // when this request is made
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        // return this mock
        result: {
          errors: [{ message: errorText }],
        },
      },
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <SingleItem id="123" />
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    const item = wrapper.find('[data-test="graphql-error"]')
    expect(item.text()).toContain(errorText)
    expect(toJSON(item)).toMatchSnapshot()
  })
})
