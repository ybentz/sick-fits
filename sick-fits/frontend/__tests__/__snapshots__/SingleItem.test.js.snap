// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem/> should render 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItem/> should render 2`] = `
<img
  alt="Item image"
  src="dog.jpg"
/>
`;

exports[`<SingleItem/> should render 3`] = `
<p>
  dogs
</p>
`;

exports[`<SingleItem/> should render error 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items not found
</p>
`;
